src/typecheck/core.ml

```diff
@@ -29,17 +34,54 @@ module Ae = struct
       | Type.Builtin Ast.False ->
         `Term (Base.app0 (module Type) env s T._false)
 
-      (* Terms connectors *)
+      (* Boolean operators *)
+      | Type.Builtin Ast.Not ->
+        `Term (Base.term_app1 (module Type) env s T.neg)
+      | Type.Builtin Ast.And ->
+        `Term (Base.term_app_left (module Type) env s mk_and)
+      | Type.Builtin Ast.Or ->
+        `Term (Base.term_app_left (module Type) env s mk_or)
+      | Type.Builtin Ast.Xor ->
+        `Term (Base.term_app_left (module Type) env s T.xor)
+      | Type.Builtin Ast.Imply ->
+        `Term (Base.term_app_right (module Type) env s T.imply)
+      | Type.Builtin Ast.Equiv ->
+        `Term (Base.term_app_right (module Type) env s T.equiv)
+
+      (* If-then-else *)
+      | Type.Builtin Ast.Ite ->
+        `Term (
+          Base.make_op3 (module Type) env s (fun _ (a, b, c) ->
+              let cond = Type.parse_prop env a in
+              let then_ = Type.parse_term env b in
+              let else_ = Type.parse_term env c in
+              T.ite cond then_ else_
+          )
+        )
+
+      (* Equality *)
       | Type.Builtin Ast.Eq ->
         `Term (Base.term_app2 (module Type) env s T.eq)
+      | Type.Builtin Ast.Distinct ->
+        `Term (
```

Comment:
Why not simply do a `Base.term_app_list` with a `T.distinct` here, as is done for e.e.g smtlib ? (Side note: maybe we need to add a parameter to `term_app_list` to specify a minimum number of arguments ?)

Reply:
Done. (I think that the check function in ```Base.term_app_list``` should be of the same type as the one in ```Base.term_app_left``` that way the caller can define whatever checks he wants to do on the ast and args)